Source/items.cpp

```diff
@@ -4519,79 +4519,69 @@ void SpawnPremium(int pnum)
 void SpawnWitch(int lvl)
 {
 	constexpr int PinnedItemCount = 3;
+	constexpr std::array<int, PinnedItemCount> PinnedItemTypes = { IDI_MANA, IDI_FULLMANA, IDI_PORTAL };
+	constexpr int MaxPinnedBookCount = 4;
+	constexpr std::array<int, MaxPinnedBookCount> PinnedBookTypes = { 114, 115, 116, 117 };
+
+	int bookCount = 0;
+	const int pinnedBookCount = gbIsHellfire ? GenerateRnd(MaxPinnedBookCount) : 0;
+	const int reservedItems = gbIsHellfire ? 10 : 17;
+	const int itemCount = GenerateRnd(WITCH_ITEMS - reservedItems) + 10;
+	const int maxValue = gbIsHellfire ? 200000 : 140000;
+
+	for (int i = 0; i < WITCH_ITEMS; i++) {
+		Item &item = witchitem[i];
+		item = {};
+
+		if (i < PinnedItemCount) {
+			item._iSeed = AdvanceRndSeed();
+			GetItemAttrs(item, PinnedItemTypes[i], 1);
+			item._iCreateInfo = lvl;
+			item._iStatFlag = true;
+			continue;
+		}
 
-	int j = PinnedItemCount;
-
-	memset(&Items[0], 0, sizeof(*Items));
-	GetItemAttrs(Items[0], IDI_MANA, 1);
-	witchitem[0] = Items[0];
-	witchitem[0]._iCreateInfo = lvl;
-	witchitem[0]._iStatFlag = true;
-	memset(&Items[0], 0, sizeof(*Items));
-	GetItemAttrs(Items[0], IDI_FULLMANA, 1);
-	witchitem[1] = Items[0];
-	witchitem[1]._iCreateInfo = lvl;
-	witchitem[1]._iStatFlag = true;
-	memset(&Items[0], 0, sizeof(*Items));
-	GetItemAttrs(Items[0], IDI_PORTAL, 1);
-	witchitem[2] = Items[0];
-	witchitem[2]._iCreateInfo = lvl;
-	witchitem[2]._iStatFlag = true;
-
-	int maxValue = 140000;
-	int reservedItems = 12;
-	if (gbIsHellfire) {
-		maxValue = 200000;
-		reservedItems = 10;
-
-		int books = GenerateRnd(4);
-		for (int i = 114, bCnt = 0; i <= 117 && bCnt < books; ++i) {
-			if (!WitchItemOk(i))
-				continue;
-			if (lvl < AllItemsList[i].iMinMLvl)
+		if (gbIsHellfire && i < PinnedItemCount + MaxPinnedBookCount && bookCount < pinnedBookCount) {
+			int bookType = PinnedBookTypes[i - PinnedItemCount];
+			if (lvl < AllItemsList[bookType].iMinMLvl)

```

Comment:
This function gets called with a `lvl` value clamped to the range 6, 16. The fourth book has a `iMinMLvl` of 20, so will never spawn.

With the way this condition is structured this PR will end up generating fewer items for low level characters (leaving a garbage in slots 4, 5, 6). Continuing the loop here leaves the current item as a zero-initialised `ItemType::Misc` instead of setting it to `ItemType::None` or generating a random item.

![image](https://user-images.githubusercontent.com/357684/143399259-6f9df585-cf29-4e9e-a437-da04a9b0a0c9.png)

Reply:
Should be fixed now